webrender/src/freelist.rs

```diff
@@ -2,104 +2,105 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
-pub struct FreeListItemId(u32);
+use std::marker::PhantomData;
+use std::mem;
 
-impl FreeListItemId {
-    #[inline]
-    pub fn new(value: u32) -> FreeListItemId {
-        FreeListItemId(value)
-    }
+// TODO(gw): Add a weak free list handle. This is like a strong
+//           free list handle below, but will contain an epoch
+//           field. Weak handles will use a get_opt style API
+//           which returns an Option<T> instead of T.
 
-    #[inline]
-    pub fn value(&self) -> u32 {
-        self.0
-    }
-}
+// TODO(gw): Add an occupied list head, for fast
+//           iteration of the occupied list to implement
+//           retain() style functionality.
 
-pub trait FreeListItem {
-    fn take(&mut self) -> Self;
-    fn next_free_id(&self) -> Option<FreeListItemId>;
-    fn set_next_free_id(&mut self, id: Option<FreeListItemId>);
+#[derive(Debug)]
+pub struct FreeListHandle<T> {
+    index: u32,
+    _marker: PhantomData<T>,
 }
 
-struct FreeListIter<'a, T: 'a> {
-    items: &'a [T],
-    cur_index: Option<FreeListItemId>,
+enum SlotValue<T> {
```

Comment:
In the future (when we have weak handles) the ```Free``` enum type will have some information as well (the previous epoch used in that slot).